### PR TITLE
#688 Removing pydantic

### PR DESCRIPTION
**Pull request by CharlesRendle**
Satisfies ticket #688

Removes the last remnants of pydantic validation (package, `PydanticModel` base class, validation functions and package dependency from csvcubed.

All tests affected by the removal have been altered or removed where our own validation functions are now tested.

### CODE REVIEW

File: external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md [hunk 1]
```diff
@@ -1,11 +0,0 @@
-# Error - unknown pydantic validation error
```
Comment (robons) [inline external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md:1]: Please don't delete this, but can you make it clear that if the user sees this error it is deprecated and they should update to the latest version of csvcubed.

We need to make sure that people using old versions of csvcubed can still access these help/error pages.

File: src/csvcubed/models/cube/qb/catalog.py [hunk 2]
```diff
@@ -3,25 +3,17 @@
...
-from csvcubed.models.validatedmodel import ValidatedModel, ValidationFunction
-from csvcubed.models.validationerror import ValidateModelPropertiesError
+from csvcubed.models.validatedmodel import ValidationFunction
 from csvcubed.utils import validations as v
-from csvcubed.utils.validators.uri import validate_uri as pydantic_validate_uri
-from csvcubed.utils.validators.uri import (
-    validate_uris_in_list as pydantic_validate_uris_in_list,
-)
-
-_logger = logging.getLogger(__name__)
```
Comment (robons) [inline src/csvcubed/models/cube/qb/catalog.py:24]: Why's the logger disappearsed? Isn't it used?

File: src/csvcubed/models/validationerror.py [hunk 3]
```diff
@@ -26,51 +27,75 @@ class SpecificValidationError(ValidationError, HasErrorUrl, ABC):
...
-class PydanticValidationError(ValidationError, ABC):
-    """
-    The error type returned from pydantic validation functionality.
-    """
+class ValidateModelPropertiesError(ValidationError):
 
-    path: List[str] = field(init=False, default_factory=lambda: [])
     """
-    Place to for pydantic validation to set the object path where this error occurred.
+    This error will be returned with the name of the variable that has not been validated.
```
Comment (robons) [inline src/csvcubed/models/validationerror.py:33]: It's not the name anymore, it's the path to the offending value.

File: src/csvcubed/utils/validations.py [hunk 4]
```diff
@@ -226,6 +226,26 @@ def validate(
     return validate
 
 
+def all_of(*conditions: ValidationFunction) -> ValidationFunction:
+    """
+    This function will validate if all arguments provided are instances of the types
```
Comment (robons) [inline src/csvcubed/utils/validations.py:231]: Not 100% sure about this comment. It validates that a property matches all of the conditions provided.

File: tests/test-cases/readers/cube-config/v1.0/cube_datatypes.json [hunk 5]
```diff
@@ -118,8 +118,13 @@
             "label": "time attribute",
             "data_type": "time"
         },
+        "Attr-number": {
+            "type": "attribute",
+            "label": "number attribute",
+            "data_type": "double"
```
Comment (robons) [inline tests/test-cases/readers/cube-config/v1.0/cube_datatypes.json:124]: Wasn't the data type here supposed to be `number`? I thought that was the whole point of the test?

File: tests/unit/cube/qb/test_qb_unit.py [hunk 6]
```diff
@@ -19,35 +19,6 @@
 SCHEMA_PATH_FILE = APP_ROOT_DIR_PATH / "schema" / "cube-config" / "v1_0" / "schema.json"
 
 
-def test_new_unit_base_unit_validation():
```
Comment (robons) [inline tests/unit/cube/qb/test_qb_unit.py:22]: Have these tests been replaced?

File: tests/unit/utils/test_validations.py [hunk 7]
```diff
@@ -936,5 +953,25 @@ def test_whole_object_in_parent_class():
     assert errors[0].offending_value == test_instance.test_unit_whole_obj
 
 
+def test_validate_all_of_correct():
+    """fill this in"""
```
Comment (robons) [inline tests/unit/utils/test_validations.py:957]: todo

File: tests/unit/utils/test_validations.py [hunk 8]
```diff
@@ -936,5 +953,25 @@ def test_whole_object_in_parent_class():
...
+
+    test_instance = TestClass(test_upper_case_string="STRING")
+
+    errors = test_instance.validate()
+
+    assert not any(errors)
+
+
+def test_validate_all_of_fail():
+    """fill this in"""
```
Comment (robons) [inline tests/unit/utils/test_validations.py:967]: todo

File: tests/unit/utils/test_validations.py [hunk 9]
```diff
@@ -936,5 +953,25 @@ def test_whole_object_in_parent_class():
...
+
+    test_instance = TestClass(test_upper_case_string="STRING")
+
+    errors = test_instance.validate()
+
+    assert not any(errors)
+
+
+def test_validate_all_of_fail():
```
Comment (robons) [inline tests/unit/utils/test_validations.py:966]: We also need a test which checks that it fails if it isn't a string.

**Review by robons — COMMENTED** [review]
Great work. I think it's pretty close to being ready.

File: external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md (hunk 1 above)
Comment (nimshi89) [reply external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md:1]: solved

File: src/csvcubed/models/validationerror.py (hunk 3 above)
Comment (nimshi89) [reply src/csvcubed/models/validationerror.py:33]: done

File: src/csvcubed/utils/validations.py (hunk 4 above)
Comment (nimshi89) [reply src/csvcubed/utils/validations.py:231]: done

File: tests/unit/utils/test_validations.py (hunk 7 above)
Comment (nimshi89) [reply tests/unit/utils/test_validations.py:957]: done

File: tests/unit/utils/test_validations.py (hunk 8 above)
Comment (nimshi89) [reply tests/unit/utils/test_validations.py:967]: done

File: tests/unit/cube/qb/test_qb_unit.py (hunk 6 above)
Comment (CharlesRendle) [reply tests/unit/cube/qb/test_qb_unit.py:22]: https://github.com/GSS-Cogs/csvcubed/blob/9b259fcf2f2b4468c0e90c0ed528b2b81c186487/tests/unit/utils/test_validations.py#L763-L805

I think they were included in `test_validations.py`

File: tests/test-cases/readers/cube-config/v1.0/cube_datatypes.json (hunk 5 above)
Comment (CharlesRendle) [reply tests/test-cases/readers/cube-config/v1.0/cube_datatypes.json:124]: Yeah it is. Have changed

File: src/csvcubed/models/cube/qb/catalog.py (hunk 2 above)
Comment (robons) [reply src/csvcubed/models/cube/qb/catalog.py:24]: If it's not referenced, it isn't doing anything so it isn't necessary. I'm just surprised it was in there if it wasn't necessary.

File: external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md (hunk 1 above)
Comment (robons) [reply external-docs/docs/guides/errors/build-command-errors/unknown-pydantic-error.md:1]: Not quite, this still deletes the title of the document.

**Review by robons — COMMENTED** [review]
oops

**Review by robons — APPROVED** [review]

**Pull request merged**